File: modules/workflow_engine.py
# modules/workflow_engine.py
# Orchestrates the full image -> video pipeline using Playwright.

import time
import logging
from playwright.sync_api import sync_playwright, BrowserContext, Page
from config import USER_DATA_DIR, HEADLESS, SLOW_MO_MS, RETRY_DELAY_SEC
from .prompt_loader   import PromptLoader
from .status_tracker  import StatusTracker
from .image_generator import ImageGenerator
from .video_generator import VideoGenerator
from .downloader      import Downloader

logger = logging.getLogger("bot")


class WorkflowEngine:
    """
    Main orchestrator.  Call run() to start the full pipeline.
    It will:
      - Launch one persistent browser session (reuses saved login)
      - Open two tabs (Google Flow + Meta AI)
      - Process every pending prompt
      - Track success / failure with automatic retry
    """

    def __init__(self):
        self.context:     BrowserContext = None
        self.flow_page:   Page = None
        self.meta_page:   Page = None
        self.loader       = PromptLoader()
        self.tracker      = StatusTracker()

    # ------------------------------------------------------------------
    # Browser lifecycle
    # ------------------------------------------------------------------
    def _start_browser(self, playwright):
        """Launch a persistent browser context that reuses saved login session."""
        self.context = playwright.chromium.launch_persistent_context(
            user_data_dir=USER_DATA_DIR,
            headless=HEADLESS,
            slow_mo=SLOW_MO_MS,
            accept_downloads=True,
            viewport={"width": 1280, "height": 800},
        )
        # Open dedicated tabs for each service
        self.flow_page = self.context.new_page()
        self.meta_page = self.context.new_page()
        logger.info("Browser launched with persistent session (user-data/)")

    def _stop_browser(self):
        """Close the browser context cleanly."""
        if self.context:
            self.context.close()
            logger.info("Browser closed")

    # ------------------------------------------------------------------
    # Step runners with retry
    # ------------------------------------------------------------------
    def _run_image_step(self, prompt: dict):
        """Run image generation with retry on failure."""
        pid          = prompt["id"]
        image_prompt = prompt["image_prompt"]
        generator    = ImageGenerator(self.flow_page)

        self.tracker.set_image_in_progress(pid)
        try:
            image_path = generator.generate(pid, image_prompt)
            self.tracker.set_image_success(pid, image_path)
            return image_path
        except Exception as e:
            error_msg = str(e)
            logger.error(f"[ID:{pid}] Image generation error: {error_msg}")
            self.tracker.set_image_failed(pid, error_msg)
            raise

    def _run_video_step(self, prompt: dict, image_path: str):
        """Run video generation with retry on failure."""
        pid              = prompt["id"]
        animation_prompt = prompt["animation_prompt"]
        generator        = VideoGenerator(self.meta_page)

        self.tracker.set_video_in_progress(pid)
        try:
            video_path = generator.generate(pid, image_path, animation_prompt)
            self.tracker.set_video_success(pid, video_path)
            return video_path
        except Exception as e:
            error_msg = str(e)
            logger.error(f"[ID:{pid}] Video generation error: {error_msg}")
            self.tracker.set_video_failed(pid, error_msg)
            raise

    # ------------------------------------------------------------------
    # Main pipeline
    # ------------------------------------------------------------------
    def run(self):
        """Entry point.  Runs the full pipeline for all pending prompts."""
        logger.info("=" * 60)
        logger.info("IMAGE-VIDEO AUTOMATION BOT  |  Starting...")
        logger.info("=" * 60)

        with sync_playwright() as playwright:
            self._start_browser(playwright)
            try:
                self._process_all()
            finally:
                self._stop_browser()

        self.tracker.print_summary()

    def _process_all(self):
        """Process image then video steps for all applicable prompts."""
        # --- Phase 1: Image generation ---
        image_prompts = self.loader.get_pending_prompts()
        if image_prompts:
            logger.info(f"Phase 1: {len(image_prompts)} image(s) to generate")
            for prompt in image_prompts:
                pid = prompt["id"]
                try:
                    self._run_image_step(prompt)
                except Exception:
                    logger.warning(f"[ID:{pid}] Skipping video step (image failed)")
                    continue
                time.sleep(RETRY_DELAY_SEC)  # Brief pause between prompts
        else:
            logger.info("Phase 1: No pending image tasks")

        # --- Phase 2: Video generation ---
        video_prompts = self.loader.get_video_pending_prompts()
        if video_prompts:
            logger.info(f"Phase 2: {len(video_prompts)} video(s) to generate")
            for prompt in video_prompts:
                pid        = prompt["id"]
                image_path = prompt.get("image_path") or Downloader.get_image_path(pid)
                try:
                    self._run_video_step(prompt, image_path)
                except Exception:
                    logger.warning(f"[ID:{pid}] Video step failed, moving to next")
                    continue
                time.sleep(RETRY_DELAY_SEC)
        else:
            logger.info("Phase 2: No pending video tasks")

File: src/workflowEngine.ts
// src/workflowEngine.ts
// Orchestrates the full image -> video pipeline using Playwright.

import { chromium } from 'playwright-extra';
import { BrowserContext, Page } from 'playwright';
import stealth from 'puppeteer-extra-plugin-stealth';


// Use stealth plugin
chromium.use(stealth());

import { USER_DATA_DIR, HEADLESS, SLOW_MO_MS, RETRY_DELAY_MS, BROWSER_CHANNEL } from './config';
import { PromptLoader, Prompt } from './promptLoader';
import { StatusTracker } from './statusTracker';
import { ImageGenerator } from './imageGenerator';
import { VideoGenerator } from './videoGenerator';
import { Downloader } from './downloader';
import logger from './logger';

export class WorkflowEngine {
  private context: BrowserContext | null = null;
  private flowPage: Page | null = null;
  private metaPage: Page | null = null;
  private loader = new PromptLoader();
  private tracker = new StatusTracker();
  private sessionCounter = 0;
  private readonly PROMPTS_BEFORE_BREAK = 5;


  private async startBrowser(): Promise<void> {
    this.context = await chromium.launchPersistentContext(USER_DATA_DIR, {
      headless: HEADLESS,
      slowMo: SLOW_MO_MS,
      acceptDownloads: true,
      channel: BROWSER_CHANNEL,
      ignoreDefaultArgs: ['--enable-automation'],
      args: [
        '--disable-blink-features=AutomationControlled',
        '--no-sandbox',
        '--disable-setuid-sandbox'
      ]
    });


    // Hardening: Inject deep stealth scripts via string to bypass TS compiler
    await this.context.addInitScript(`
      // 1. Hide webdriver
      Object.defineProperty(navigator, 'webdriver', { get: () => undefined });

      // 2. Fake chrome.runtime
      window.chrome = { runtime: {}, loadTimes: () => {}, csi: () => {} };

      // 3. Spoof plugins & mimeTypes
      Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3, 4, 5] });
      Object.defineProperty(navigator, 'mimeTypes', { get: () => [1, 2, 3, 4, 5] });

      // 4. Override permissions query
      const originalQuery = window.navigator.permissions.query;
      window.navigator.permissions.query = (parameters) =>
        parameters.name === 'notifications'
          ? Promise.resolve({ state: Notification.permission })
          : originalQuery(parameters);

      // 5. Spoof WebGL renderer (Intel Iris)
      const getParameter = WebGLRenderingContext.prototype.getParameter;
      WebGLRenderingContext.prototype.getParameter = function(parameter) {
        if (parameter === 37445) return 'Intel Inc.';
        if (parameter === 37446) return 'Intel Iris OpenGL Engine';
        return getParameter.call(this, parameter);
      };
    `);


    // Reuse the first tab that Chrome opens automatically
    const pages = this.context.pages();
    this.flowPage = pages.length > 0 ? pages[0] : await this.context.newPage();

    // Open a second tab for Meta (even though we have it commented out, this keeps the structure)
    this.metaPage = await this.context.newPage();

    logger.info('Browser launched with stealth hardening.');
  }



  private async stopBrowser(): Promise<void> {
    if (this.context) {
      await this.context.close();
      logger.info('Browser closed');
    }
  }

  private async runImageStep(prompt: Prompt): Promise<string> {
    const { id, image_prompt } = prompt;
    const generator = new ImageGenerator(this.flowPage!);
    this.tracker.setImageInProgress(id);
    try {
      const imagePath = await generator.generate(id, image_prompt);
      this.tracker.setImageSuccess(id, imagePath);
      return imagePath;
    } catch (err) {
      const msg = String(err);
      logger.error(`[ID:${id}] Image error: ${msg}`);
      this.tracker.setImageFailed(id, msg);
      throw err;
    }
  }

  private async runVideoStep(prompt: Prompt, imagePath: string): Promise<void> {
    const { id, animation_prompt } = prompt;
    const generator = new VideoGenerator(this.metaPage!);
    this.tracker.setVideoInProgress(id);
    try {
      const videoPath = await generator.generate(id, imagePath, animation_prompt);
      this.tracker.setVideoSuccess(id, videoPath);
    } catch (err) {
      const msg = String(err);
      logger.error(`[ID:${id}] Video error: ${msg}`);
      this.tracker.setVideoFailed(id, msg);
      throw err;
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  private async processAll(): Promise<void> {
    // Phase 1: Image generation
    const imagePrompts = this.loader.getPendingImagePrompts();
    if (imagePrompts.length) {
      logger.info(`Phase 1: ${imagePrompts.length} image(s) to generate`);
      for (const prompt of imagePrompts) {
        try {
          await this.runImageStep(prompt);
        } catch {
          logger.warn(`[ID:${prompt.id}] Skipping video (image failed)`);
          continue;
        }

        // --- Session Rate Limiting ---
        this.sessionCounter++;
        if (this.sessionCounter >= this.PROMPTS_BEFORE_BREAK) {
          const breakMinutes = Math.floor(Math.random() * 3) + 3; // 3-5 mins
          logger.info(`[SESSION] Reached ${this.PROMPTS_BEFORE_BREAK} prompts. Taking a human break for ${breakMinutes} minutes...`);
          await this.sleep(breakMinutes * 60 * 1000);
          this.sessionCounter = 0;
          logger.info('[SESSION] Break over. Resuming automation...');
        }

        // Hardening: Random delay between prompts (10-25s)
        const delay = Math.floor(Math.random() * 15_000) + 10_000;
        logger.info(`Waiting ${delay / 1000}s before next prompt to remain stealthy...`);
        await this.sleep(delay);

      }

    } else {
      logger.info('Phase 1: No pending image tasks');
    }

    // Phase 2: Video generation
    const videoPrompts = this.loader.getPendingVideoPrompts();
    if (videoPrompts.length) {
      logger.info(`Phase 2: ${videoPrompts.length} video(s) to generate`);
      for (const prompt of videoPrompts) {
        const imagePath = prompt.image_path || Downloader.getImagePath(prompt.id);
        try {
          await this.runVideoStep(prompt, imagePath);

          // --- Session Rate Limiting ---
          this.sessionCounter++;
          if (this.sessionCounter >= this.PROMPTS_BEFORE_BREAK) {
            const breakMinutes = Math.floor(Math.random() * 3) + 3; // 3-5 mins
            logger.info(`[SESSION] Reached ${this.PROMPTS_BEFORE_BREAK} prompts. Taking a human break for ${breakMinutes} minutes...`);
            await this.sleep(breakMinutes * 60 * 1000);
            this.sessionCounter = 0;
            logger.info('[SESSION] Break over. Resuming automation...');
          }

          // Hardening: Random delay between prompts (10-25s)
          const delay = Math.floor(Math.random() * 15_000) + 10_000;
          logger.info(`Waiting ${delay / 1000}s before next video prompt to remain stealthy...`);
          await this.sleep(delay);
        } catch {
          logger.warn(`[ID:${prompt.id}] Video step failed, moving to next`);
          await this.sleep(RETRY_DELAY_MS);
        }

      }
    } else {
      logger.info('Phase 2: No pending video tasks');
    }


  }

  async run(): Promise<void> {
    logger.info('='.repeat(60));
    logger.info('IMAGE-VIDEO AUTOMATION BOT  |  Starting...');
    logger.info('='.repeat(60));
    await this.startBrowser();
    try {
      await this.processAll();
    } finally {
      await this.stopBrowser();
    }
    this.tracker.printSummary();
  }
}
